server/hooks.go

```diff
@@ -17,26 +17,65 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
 		return
 	}
 
-	for _, message := range p.getWelcomeMessages() {
-		if message.TeamName == data.Team.Name {
-			go p.processWelcomeMessage(*data, *message)
+	key := fmt.Sprintf("%s%s", welcomebotTeamWelcomeKey, teamMember.TeamId)
+	teamMessage, appErr := p.API.KVGet(key)
+	if appErr != nil {
+		mlog.Error(
+			"error occurred while retrieving the welcome message",
+			mlog.String("teamId", teamMember.TeamId),
+			mlog.Err(appErr),
+		)
+		return
+	}
+
+	if teamMessage == nil {
+		// No dynamic welcome message for the given team, so we check if one as been set in the config.json
+		for _, message := range p.getWelcomeMessages() {
+			if message.TeamName == data.Team.Name {
+				go p.processWelcomeMessage(*data, *message)
+			}
 		}
+		return
+	}
+
+	// We send a DM and an opportunistic ephemeral message to the channel. See
+	// the discussion at the link below for more details:
+	// https://github.com/mattermost/mattermost-plugin-welcomebot/pull/31#issuecomment-611691023
+	teamAddMessage := "# Welcome to the " + data.Team.DisplayName + " team!! \n\n"
+	postDM := &model.Post{
+		UserId:    p.botUserID,
+		ChannelId: data.DirectMessage.Id,
+		Message:   teamAddMessage + string(teamMessage),
+	}
+	if _, appErr := p.API.CreatePost(postDM); appErr != nil {
+		mlog.Error("failed to post welcome message to the channel",
+			mlog.String("channelId", data.DirectMessage.Id),
+			mlog.Err(appErr),
+		)
 	}
+
+	postChannel := &model.Post{
+		UserId:    p.botUserID,
+		ChannelId: teamMember.TeamId,
+		Message:   string(teamMessage),
+	}
+	time.Sleep(1 * time.Second)
+	_ = p.API.SendEphemeralPost(teamMember.UserId, postChannel)
```

Comment:
There is a lot of duplicate code here. Can this be avoided?

Reply:
I can go through this piece and refactor it to be less complicated!

Reply:
I removed the ephemeral message. I believe we originally had it so that the user would immediately be greeted by the welcome message, however the DM with the welcome-bot is plenty sufficient.